### PR TITLE
Add event system to stateful collection

### PR DESCRIPTION
**Pull request by simplicbe**
I've checked the following points:

- [x] I've formatted the code by pressing (STRG + K + D)
- [x] I've removed unnecessary usings
- [x] I've not used magic variables
- [x] I've used localization service instead of text
- [x] All `public` things have comments
- [x] This has been followed: [conventions](https://simplic.github.io/dev/csharp/getting_started/conventions.html)

### CODE REVIEW

File: simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs [hunk 1]
```diff
@@ -1,275 +1,409 @@
...
+            if (removedItems.Contains(item))
+            {
+                var args = new AddItemEventArgs
+                {
+                    Item = item,
+                    ReaddFromRemovedItems = true
+                };
+                AddItem?.Invoke(this, args);
+
+                if (!args.Cancel)
```
Comment (yvz5) [inline simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs]: do you really need to check this ? you are creating the args yourself. you know which value the args has.

File: simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs [hunk 2]
```diff
@@ -1,275 +1,409 @@
...
+            var args = new ClearCollectionEventArgs();
+            ClearCollection?.Invoke(this, args);
+
+            if (!args.Cancel)
+            {
+                items.Clear();
+                newItems.Clear();
+                removedItems.Clear();
+
+                var collectionClearedArgs = new CollectionClearedEventArgs();
```
Comment (yvz5) [inline simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs]: maybe put counters in the args: ItemsCleared: 45, NewItemsCleared: 2, RemovedItemsCleared: 3

File: simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs [hunk 3]
```diff
@@ -1,275 +1,409 @@
...
+        }
+
+        /// <summary>
+        /// Check whether an items exists in the items list or the new items list
+        /// </summary>
+        /// <param name="item">Item to proof</param>
+        /// <returns>True if the item exists</returns>
+        public bool Contains(T item)
+        {
+            if (items.Contains(item) || newItems.Contains(item))
```
Comment (yvz5) [inline simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs]: `return (items.Contains(item) || newItems.Contains(item))` is way cleaner imho

File: simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs [hunk 4]
```diff
@@ -1,275 +1,409 @@
...
+
+            return false;
+        }
+
+        /// <summary>
+        /// Copy current items into an array
+        /// </summary>
+        /// <param name="array">Array to copy items to</param>
+        /// <param name="arrayIndex">Start index</param>
+        public void CopyTo(T[] array, int arrayIndex)
```
Comment (yvz5) [inline simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs]: I think a method such as `ToArray()` is way better. but i assume this is a legacy method

File: simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs [hunk 5]
```diff
@@ -1,275 +1,409 @@
...
+            get
+            {
+                return removedItems.Count;
+            }
+        }
+
+        /// <summary>
+        /// Get the amount of new items
+        /// </summary>
+        public int CountNewItems
```
Comment (yvz5) [inline simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs]: i would put these into event args as i mentioned above

File: simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs (hunk 1 above)
Comment (simplicbe) [reply simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs]: What do you mean? I check whether the event is not null, Böttcher args...

Comment (yvz5) [reply simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs]: > What do you mean? I check whether the event is not null, **Böttcher** args...

iphone plays games again 😆 

i meant `if (!args.Cancel)`

File: simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs (hunk 4 above)
Comment (yvz5) [reply simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs]: possible index out of range exception. you should check the `arrayIndex` if its inside the arrays size. To be honest, this method does not belong here. its not this objects job to make sure array is not empty, index is within the range and other possible bugs. Just return a new array

**Pull request merged**